lib/memjs/memjs.js

```diff
@@ -170,6 +186,137 @@ Client.prototype.get = function(key, callback) {
   });
 };
 
+/** Build a pipelined get multi request by sending one GETKQ for each key (quiet, meaning it won't respond if the value is missing) followed by a no-op to force a response (and to give us a sentinel response that the pipeline is done)
+ *
+ * cf https://github.com/couchbase/memcached/blob/master/docs/BinaryProtocol.md#0x0d-getkq-get-with-key-quietly
+ */
+Client.prototype._buildGetMultiRequest = function(keys) {
+  // start at 24 for the no-op command at the end
+  var requestSize = 24;
+  for (var keyIdx in keys) {
+    requestSize += Buffer.byteLength(keys[keyIdx], 'utf8') + 24;
+  }
+
+  var request = Buffer.alloc(requestSize);
+  request.fill();
+
+  this.incrSeq();
+
+  var bytesWritten = 0;
+  for (keyIdx in keys) {
+    var key = keys[keyIdx];
+    bytesWritten += addToRequestBuffer(constants.OP_GETKQ, key, '', '', this.seq, request, bytesWritten);
+  }
+
+  bytesWritten += addToRequestBuffer(
+    constants.OP_NO_OP, '', '', '', this.seq, request, bytesWritten);
+
+  return request;
+};
+
+/** Executing a pipelined (multi) get against a single server. This is a private implementation detail of getMulti. */
+Client.prototype._getMultiToServer = function (serv, keys, callback) {
+  var self = this;
+
+  var responseMap = {};
+
+  var handle = function(response) {
+    switch (response.header.status) {
+    case  0:
+      if (callback) {
+        var deserialized = self.serializer.deserialize(response.header.opcode, response.val, response.extras);
+        // When we get the no-op response, we are done with this one getMulti in the per-backend fan-out
+        if (response.header.opcode === constants.OP_NO_OP) {
+          // this ensures the handler will be deleted from the responseCallbacks map in server.js
+          handle.quiet = true;
+          callback(null, responseMap);
+        } else {
+          var key = response.key.toString();
+          responseMap[key] = deserialized.value;
+        }
+      }
+      break;
+    case 1:
+      if (callback) { callback(null, null, null); }
```

Comment:
What is this case?

Reply:
Do we need to set `handle.quiet = true` here so that the handler gets GC'd?

Reply:
Okay, after thinking a bit more about this - we only want to set `handle.quiet = false` so the handler gets deleted once we get to the no-op response. But if the no-op response is an error (default case) or whatever this is (???), then we should still make sure to remove the handler.

Reply:
changed all the case statements in here to use a constant to make this somewhat clearer.

I think you're right about handle.quiet = false - also will add tests

Reply:
Does this get called only once for the NO_OP, or can it get called multiple times for getkq responses after the server gets the no-op?

Reply:
my understanding from the docs is that memcache always processes and responds to commands in order - so for this sequence number (remember callbacks are tied to one seq num) nothing should come back after the no-op

from the docs:

"The current protocol dictates that the server won't start processing the next command until the current command is completely processed (due to the lack of barriers or any other primitives to enforce execution order). The protocol defines some "quiet commands" which won't send responses in certain cases (success for mutations, not found for gets etc). The client would know that such commands was executed when it encounters the response for the next command requested issued by the client."